google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageReadChannel.java

```diff
@@ -101,26 +105,33 @@ public int read(ByteBuffer dst) throws IOException {
       checkOpen();
       int amt;
       int retries = 0;
-      int maxRetries = 3;
+      int maxRetries = Math.max(3, maxChannelReopens);
       dst.mark();
       while (true) {
         try {
           dst.reset();
           amt = channel.read(dst);
           break;
         } catch (StorageException exs) {
-          if (exs.getMessage().contains("Connection closed prematurely") && reopens < maxChannelReopens) {
-            // this error isn't marked as retryable since the channel is closed;
-            // but here at this higher level we can retry it.
+          if (isReopenable(exs)) {
+            // these errors aren't marked as retryable since the channel is closed;
+            // but here at this higher level we can retry them.
             reopens++;
+            if (reopens > maxChannelReopens) {
+              throw new StorageException(exs.getCode(), "All reopens failed", exs);
+            }
             sleepForAttempt(reopens);
             innerOpen();
             continue;
-          } else if ((exs.getCode() == 500 || exs.getCode() == 503)  && retries < maxRetries) {
+          } else if (exs.isRetryable() || exs.getCode() == 500 || exs.getCode() == 503) {
             retries++;
+            if (retries > maxRetries) {
+              throw new StorageException(exs.getCode(), "All retries failed", exs);
```

Comment:
Ok, it is a new change... In most cases, this new StorageException will be initialized as retriable (.isRetryable() will return true, and currently there is no way to set retriable property explicitly, it is set based on the error code). It can be confusing for users (i.e. a code, after failed retries throwing an exception which says that it is still "retriabel == true"). Maybe at least document this. 

Also (not related to your change), isRetryable seems like a misspelling from our side (though I'm not even sure than either "retryable" or "retriable" exist in English dictionary at all=)